v2/internal/reconcilers/entra/entra_securitygroup_reconciler.go

```diff
@@ -0,0 +1,499 @@
+// ctx is the context for the operation.
+// id is the Entra ID of the group to update.
+// group is the security group to update.
+// eventRecorder is used to record events for the group.
+// log is the logger to use for logging.
+// Returns the Entra ID of the group to adopt, and nil if found;
+// nil, nil if no group was found to adopt;
+// or nil, error if there was a problem finding the group.
+func (r *EntraSecurityGroupReconciler) tryAdopt(
+	ctx context.Context,
+	group *asoentra.SecurityGroup,
+	log logr.Logger,
+) (*string, error) {
+	log.V(Status).Info("Searching for existing Entra security group to adopt", "group", group.Name)
+
+	// Create our Entra Client
+	client, err := r.EntraClientFactory(ctx, group)
+	if err != nil {
+		return nil, eris.Wrap(err, "creating entra client prior to adoption search")
+	}
+
+	// Try to find the group by DisplayName
+	displayName := group.Spec.DisplayName
+	if displayName == nil || *displayName == "" {
+		// Can't adopt without a display name
+		return nil, nil
+	}
+
+	log.V(Status).Info("Searching for existing Entra security group by display name", "displayName", *displayName)
+	groups, err := r.loadGroupsByDisplayName(ctx, *displayName, client.Client())
+	if err != nil {
+		if r.isNotFound(err) {
+			// No group to adopt
+			return nil, nil
+		}
+
+		return nil, eris.Wrapf(err, "getting group by display name %s", *displayName)
+	}
+
+	if len(groups) == 0 {
+		// No group to adopt
+		log.V(Status).Info("No existing Entra security group found by display name", "displayName", *displayName)
+		return nil, nil
+	}
+
+	if len(groups) > 1 {
+		// Multiple groups found with the same display name
+		log.V(Status).Info("Multiple existing Entra security groups found by display name", "displayName", *displayName)
+		return nil, eris.Errorf("multiple existing Entra security groups found with display name %s", *displayName)
+	}
+
+	// We found a single group to adopt
+	g := groups[0]
+	return g.GetId(), nil
+}
+
+// create completes our reconciliation by creating a new Entra security group.
+// ctx is the context for the operation.
+// group is the security group to create.
+// eventRecorder is used to record events for the group.
+// log is the logger to use for logging.
+func (r *EntraSecurityGroupReconciler) create(
+	ctx context.Context,
+	group *asoentra.SecurityGroup,
+	log logr.Logger,
+) (ctrl.Result, error) {
+	log.V(Status).Info("Creating Entra security group")
+
+	// Create our Entra Client
+	client, err := r.EntraClientFactory(ctx, group)
+	if err != nil {
+		return reconcile.Result{}, eris.Wrap(err, "creating entra client prior to adoption search")
+	}
+
+	g := msgraphmodels.NewGroup()
+	group.Spec.AssignToGroup(g)
+
+	status, err := client.Client().Groups().Post(ctx, g, nil)
+	if err != nil {
+		// Failed to create
+		return ctrl.Result{}, eris.Wrapf(err, "failed to create group %s", group.Name)
+	}
+
+	group.Status.AssignFromGroup(status)
+	setEntraID(group, *status.GetId())
```

Comment:
Is this ptr deref safe? Consider guarding it?

Reply:
Added guard clause.